File: dotnet/path_traversal/FileUpload.cs
using System.IO;
using System.IO.Compression;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Mvc;

[ApiController]
[Route("[controller]")]
public class FileUploadController : ControllerBase
{
    private const string UploadDir = "/var/www/uploads";

    // VULN 1: Save uploaded file to user-specified destination - arbitrary file write
    [HttpPost("upload")]
    public IActionResult UploadFile(IFormFile file, string destination)
    {
        string savePath = Path.Combine(UploadDir, destination);
        using var stream = new FileStream(savePath, FileMode.Create);
        file.CopyTo(stream);
        return Ok("uploaded");
    }

    // VULN 2: Zip extraction to user-controlled directory - zip slip vulnerability
    [HttpPost("extract")]
    public IActionResult ExtractArchive(IFormFile archive, string extractTo)
    {
        string archivePath = Path.Combine(UploadDir, archive.FileName);
        using (var stream = new FileStream(archivePath, FileMode.Create))
            archive.CopyTo(stream);

        // ZipFile.ExtractToDirectory does not validate entry paths against extractTo
        ZipFile.ExtractToDirectory(archivePath, extractTo, overwriteFiles: true);
        return Ok("extracted");
    }

    // VULN 3: Read file using user-controlled filename - path traversal on read
    [HttpGet("preview")]
    public IActionResult PreviewUpload(string user, string file)
    {
        string path = UploadDir + "/" + user + "/" + file;
        string content = System.IO.File.ReadAllText(path);
        return Ok(content);
    }
}

File: java/path_traversal/FileUpload.java
import jakarta.servlet.http.*;
import java.io.*;
import java.nio.file.*;
import java.util.zip.*;

public class FileUpload extends HttpServlet {
    private static final String UPLOAD_DIR = "/var/www/uploads";

    // VULN 1: Write uploaded file to user-specified destination - arbitrary file write
    protected void doUpload(HttpServletRequest req, HttpServletResponse resp) throws IOException, jakarta.servlet.ServletException {
        String dest = req.getParameter("destination");
        Part filePart = req.getPart("file");
        Path savePath = Paths.get(UPLOAD_DIR).resolve(dest);
        Files.copy(filePart.getInputStream(), savePath, StandardCopyOption.REPLACE_EXISTING);
        resp.getWriter().write("uploaded");
    }

    // VULN 2: Zip extraction to user-controlled directory - zip slip vulnerability
    protected void doExtract(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String extractTo = req.getParameter("extract_to");
        String archiveName = req.getParameter("archive");
        File archiveFile = new File(UPLOAD_DIR + "/" + archiveName);
        try (ZipInputStream zis = new ZipInputStream(new FileInputStream(archiveFile))) {
            ZipEntry entry;
            while ((entry = zis.getNextEntry()) != null) {
                // zip slip: entry name may contain ../ sequences
                File destFile = new File(extractTo, entry.getName());
                Files.copy(zis, destFile.toPath(), StandardCopyOption.REPLACE_EXISTING);
            }
        }
        resp.getWriter().write("extracted");
    }

    // VULN 3: Read file using user-controlled filename - path traversal on read
    protected void doPreview(HttpServletRequest req, HttpServletResponse resp) throws IOException {
        String username = req.getParameter("user");
        String filename = req.getParameter("file");
        String path = UPLOAD_DIR + "/" + username + "/" + filename;
        byte[] data = Files.readAllBytes(Paths.get(path));
        resp.getOutputStream().write(data);
    }
}
